brew-audit: move style checks into main audit output

apjanke opened the pull request:
###   Checklist   ###

- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [X] Have you successfully ran `brew tests` with your changes locally?

###   Description   ###

Breaks up the `brew style` part of the check for `brew audit --strict` and runs it separately for each formula, when that formula is being checked for other audits. This collects all violations for each formula in a single place, instead of doing `brew style` outputs for all formulae first, and then the other audit checks. Easier to read. It also makes the violation count at the end of the output include the `style` violations.

Before:

<img width="713" alt="audit-style - before" src="https://cloud.githubusercontent.com/assets/2618447/14620836/1ea365c2-058d-11e6-9750-7426a58d0411.png">

After:

<img width="713" alt="audit-style - after" src="https://cloud.githubusercontent.com/assets/2618447/14620838/22b4b814-058d-11e6-9450-5cfaa25b8f97.png">

This also localizes the `--fix` argument support in `style` so that it doesn't leak to `audit`. I'm not sure if that's the right thing to do, though. `audit` isn't documented as supporting `--fix`, but that doesn't necessarily mean it isn't supposed to do it.

###  Considerations  ###

1. This turns off the syntax highlighting for the Rubocop output. I think that's fine, since we're uniformly using the main `audit` output style here. If someone really misses it, I could add an option to enable the old behavior, with all the style checks at the top and with coloring intact. Or we could add `Tty`-based coloring to `audit`'s own output, and you'll get it everywhere. (Personally, I don't care. I think this much use of color is distracting instead of helpful.)
2. Will be slightly slower when auditing multiple formulae due to multiple Rubocop invocations. I don't think this is a big deal: the main use case is single-file formula audits, I think. And it was still perfectly tolerable when I ran it interactively on big batches like `brew audit --strict a*`.
3. Is `brew audit` actually supposed to support `--fix`?

mikemcquaid commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -230,6 +227,12 @@ def audit_file
+    return unless @style
+
+    style_violations = nil
+    Homebrew.check_style(@formula.path) { |rslt| style_violations = rslt }
```
No idea what `rslt` means 😉

xu-cheng replied:
Why not just 
```ruby
Homebrew.check_style(@formula.path) do |violations|
  violations.each { |violation| problem violation }
end

apjanke replied:
Sorry; shorthand for "result". I'm still learning to remember to spell everything out for `brew` code.

xu-cheng's version didn't occur to me since I'm still new at Ruby. Switched to that.

mikemcquaid commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -28,12 +28,8 @@ def audit
-      style
-    end
-
+    do_style = strict && ARGV.resolved_formulae.any? && RUBY_VERSION.split(".").first.to_i >= 2
```
`check_style`?

apjanke replied:
Nope... the `check_style` calls are moved way down below at the end of each individual formula audit. Running it here would either make the output separate, or give the results for many formulae, requiring additional parsing work.

mikemcquaid commented:
A few comments but otherwise LGTM 👍

xu-cheng commented on Library/Homebrew/cmd/style.rb:
```diff
@@ -8,18 +19,38 @@ def style
-    Homebrew.failed = !$?.success?
+    if block_given?
+      txt = nil
+      Utils.popen(["rubocop", *args], "r") { |io| txt = io.read }
```
`Utils.popen_read`?

apjanke replied:
`popen_read` returns the results in `BINARY` and would require a separate `force_encoding('UTF-8')` call and check for Ruby 1.9+. I thought just doing a `popen` with a `r` instead of `rb` would be simpler.

xu-cheng commented on Library/Homebrew/cmd/style.rb:
```diff
@@ -8,18 +19,38 @@ def style
+      # Last two lines are summary, first line, if any, is header
+      lines[-2..-1] = []
+      lines.delete_at(0) unless lines.empty?
+      yield lines
```
What will happen if I do `check_style([multiple-files]) { |lines| puts lines }`?

xu-cheng replied:
Rubocop supports JSON output. c.f. https://github.com/bbatsov/rubocop#formatters

I wonder should we use that instead of trying to parsing the output by ourselves.

mikemcquaid replied:
👍

apjanke replied:
> What will happen if I do check_style([multiple-files]) { |lines| puts lines }? 

You will get bogus output with extra lines including the formula name headers.

I didn't know about the Rubocop formatters. Yeah, that would be more appropriate to use here. I'll look at switching it over. Thanks!

apjanke commented:
Switched to "in progress" while I look at switching to Rubocop's JSON output instead of parsing their normal results output.

apjanke commented:
I rewrote the `rubocop` interaction using its JSON output, per @xu-cheng's suggestion. Now we're using the appropraite machine-readable format, results for multiple files are represented correctly, and the `rubocop` checks can be done in a single batch up front to avoid the cost of shelling out to `rubocop` separately for each formula.

apjanke commented:
Question: the `==> audit problems` `ohai` header is only displayed when `--strict` is given. Why is that?

mikemcquaid commented:
> Question: the ==> audit problems ohai header is only displayed when --strict is given. Why is that?

Could just be a mistake. Feel free to address it.

mikemcquaid commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -4,15 +4,18 @@
 #:    `audit` exits with a non-zero status if any errors are found. This is useful,
 #:    for instance, for implementing pre-commit hooks.
 
+# Undocumented options:
```
Looks like documentation to me 😉. Will this be enabled with `--debug`? If so I think it can just be omitted.

apjanke replied:
Surprisingly, no, because it takes a capital `D` instead of a lowercase `d` like `--debug`. I don't know why it's different, but it's a distinct option.

mikemcquaid replied:
Weird. Given I've been working on Homebrew for a long time and had no idea I think it's worth just using `ARGV.debug?` instead.

apjanke replied:
I took a closer look at it. All this one does is turn on instrumentation for profiling the `audit_*` methods. It doesn't turn on all the other verbose debugging stuff. And since it does profiling, it probably doesn't want to enable the other debugging stuff, since that would skew the timings.

I reworded the comment and factored out that bit of code to make it more understandable.

UniqMartin replied:
Just wanted to mention that [`brew doctor` has the exact same feature](https://github.com/Homebrew/brew/blob/b53f0c5ada753d5ded6f6fbe4f920a4a43696394/Library/Homebrew/cmd/doctor.rb#L16). It is equally undocumented.

apjanke replied:
I refactored this to share `brew doctor`'s implementation, and added a similar "Undocumented options" comment at the top of `brew doctor`. Are you folks okay with the comment there? That way the `-D` is still hidden in the man page, but when you're looking at the source there's something to indicate the intent, which, judging from this string of comments and my own experience, was not clear.

mikemcquaid commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -28,12 +31,7 @@ def audit
-      style
-    end
-
+    style = strict && ARGV.resolved_formulae.any? && RUBY_VERSION.split(".").first.to_i >= 2
```
Maybe split `RUBY_VERSION.split(".").first.to_i >= 2` into a separate variable (perhaps even a global constant)?

apjanke replied:
Done.

mikemcquaid commented on Library/Homebrew/cmd/style.rb:
```diff
@@ -8,18 +23,94 @@ def style
+  end
+
+  def check_style_impl(files, output_type, options = {})
+    files = [files] unless files.is_a? Array
```
Could just to `files = files.to_a`

xu-cheng replied:
Or, we could make it always accept an Array as input, which seems already to be the case.

apjanke replied:
`String` doesn't support `to_a`.

I just removed the conversion line: anything coming in is `concat`-able with arrays; the conversion guard isn't really needed with the current code.

mikemcquaid commented:
Looks good to me 👍

xu-cheng commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -1235,3 +1232,31 @@ def problem(text)
+
+# Debugging support
+
+def instrument_module_with_method_timers(the_module)
```
It would be better to turn it to a method of `FormulaAuditor` like what `brew doctor` does.

xu-cheng replied:
You could also make this method the same name as used in `brew doctor`.

xu-cheng commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -1235,3 +1232,31 @@ def problem(text)
+    end
+  end
+
+  if $times.nil?
```
This check is superfluous

UniqMartin commented on Library/Homebrew/cmd/style.rb:
```diff
@@ -1,3 +1,18 @@
+#:  * `style` [`--fix`] <targets>
```
Two issues:

1. This line needs to end in a colon (compare with other commands). Otherwise the styling as a definition list is broken and this is causing the huge diff in `brew.1.html` and `brew.1`.

2. `<targets>` are optional, i.e. make that `[<targets>]`. IMO, even better would be to make that `[<formulae>|<paths>]` and to adjust the text below accordingly.

apjanke replied:
Fixed.

xu-cheng commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -1235,3 +1232,31 @@ def problem(text)
+
+if ARGV.switch? "D"
+  instrument_module_with_method_timers(FormulaAuditor)
+end
```
This should be called inside `audit` method rather than when we load this file.

UniqMartin commented on Library/Homebrew/cmd/style.rb:
```diff
@@ -1,3 +1,18 @@
+#:  * `style` [`--fix`] <targets>
+#:    Check formulae or files
```
What are we checking for? The current wording is a bit too vague. Also misses a trailing full stop.

apjanke replied:
Fixed.

xu-cheng commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -23,65 +26,51 @@
 require "date"
 
 module Homebrew
+  RUBY_2_OR_LATER = RUBY_VERSION.split(".").first.to_i >= 2
```
This seems to be used multiple times thoroughout the codebase. We probably need a method in `utils.rb` for this.

apjanke replied:
Yeah. Same with the test for whether encodings are supported. I'd rather defer that to another PR, though.

mikemcquaid replied:
👍 to another PR.

UniqMartin commented on Library/Homebrew/cmd/style.rb:
```diff
@@ -8,18 +23,93 @@ def style
-    args << "--auto-correct" if ARGV.homebrew_developer? && ARGV.flag?("--fix")
+    case output_type
+    when :print
+      system "rubocop", "--format", "simple", *args
```
Can we add the RuboCop option `--display-cop-names` (short `-D`) here? It makes it a lot easier to look up information when researching how to resolve a style violation (or what needs to be adjusted, if changing the code is not desirable). Compare:

```
$ rubocop --format simple --force-exclusion --config Library/.rubocop.yml Library/brew.rb
== Library/brew.rb ==
W:139:  1: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
```

with

```
$ rubocop -D --format simple --force-exclusion --config Library/.rubocop.yml Library/brew.rb
== Library/brew.rb ==
W:139:  1: Lint/RescueException: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
```

apjanke replied:
Added the ability to do this, with an `--display-cop-names` option on `brew audit` and `brew style`, so people can try it both ways. Don't know if it should be the default: it's easier to identify, but it gets a little long and noisy-looking.

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -23,65 +26,51 @@
+      files = ARGV.resolved_formulae.map(&:path)
+    end
+    if style
+      require "cmd/style"
```
Let's put this `require` at the top of the file and reserve this kind of on-demand `require` for cases where it makes more sense (e.g. loading code that only works with a certain Ruby version or when the performance impact is non-negligible).

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -23,65 +26,51 @@
-      Formula
+    if ARGV.named.empty?
+      ff = Formula
+      files = [HOMEBREW_LIBRARY]
```
I'm not sure why we'd want to scan all of `Library` (including the code in `Library/Homebrew`) in this case, as this command is only about auditing formulae. I'd make that just `files = ff.map(&:path)` or if that creates a too long argument list for invoking `rubocop`, make that `Tap.map(&:formula_dir)` instead (array of formula directories of all installed taps).

Here's my proposal for this line and some of the surrounding code:

```ruby
ff = if ARGV.named.empty?
  Formula
else
  ARGV.resolved_formulae
end

if style
  # Check style in a single batch run up front for performance
  files = ff.map(&:path)
  style_results = check_style_json(files, :realpath => true)
end
```

apjanke replied:
That makes sense. This code is just preserving existing behavior for the old `brew style` call it did. Makes more sense just to do taps. I'll look at changing it, with something like your example.

apjanke replied:
You know what? Style checks weren't running anyway when no formula argument was supplied, because `resolved_formulae` was empty in that case.

```
style = strict && ARGV.resolved_formulae.any? && RUBY_2_OR_LATER
```

I think that was just a limitation of the implementation of `style` there, so I'm changing it to do style for the check-all case, too.

apjanke replied:
`Tap.map(&:formula_dir)` works great. Used that and activated the style checks.

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -23,65 +26,51 @@
+        ohai "audit problems"
+        puts
+        output_header = true
       end
```
I think we can completely eliminate the `unless output_header` block. It only existed in the past because if you passed `--strict` it would be difficult to distinguish between when the output of `brew style` ends and the output of `brew audit` begins. Now that the style violations are folded into the list of audit problems, this header is no longer necessary. Just like with `brew audit` (without `--strict`) before this PR.

apjanke replied:
Ohhhhh. That explains it. Yes, let's just remove it.

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -23,65 +26,51 @@
+
+      formula_count += 1
+      problem_count += fa.problems.size
+      puts "#{f.full_name}:", fa.problems.map { |p| " * #{p}" }, ""
```
That's probably personal preference, but I find the additional empty lines inserted here don't contribute to the readability of the output and can be omitted. The indentation and the bullets (`*`) are enough for me.

apjanke replied:
My personal preference would also be to not do the empty lines; I'm preserving additional behavior here.

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -1235,3 +1232,31 @@ def problem(text)
   end
 end
+
+# Debugging support
```
I'm not really happy with this code being in the global scope. Can we make that nicer, maybe by drawing some inspiration [from `Homebrew::Diagnostic.inject_dump_stats!`](https://github.com/Homebrew/brew/blob/b53f0c5ada753d5ded6f6fbe4f920a4a43696394/Library/Homebrew/diagnostic.rb#L88-L101)?

apjanke replied:
Yes. In fact, we can probably just call `inject_dump_stats!` since it looks like the same functionality. Thanks for pointing that out.

apjanke replied:
Okay, I refactored it to share an implementation with `Diagnostic.inject_dump_stats!`.

apjanke replied:
Oops, wait, still some globals left. Hold on...

UniqMartin commented on Library/Homebrew/diagnostic.rb:
```diff
@@ -1435,5 +1420,34 @@ def all
         methods.map(&:to_s).grep(/^check_/)
       end
     end # end class Checks
+
```
Superfluous empty line here.

UniqMartin commented on Library/Homebrew/diagnostic.rb:
```diff
@@ -1435,5 +1420,34 @@ def all
+
+  end
+
+  def inject_dump_stats!(the_module, pattern = /.*/)
```
I think I prefer this to have no default for `pattern`. Now that this method is used by multiple commands, it should definitely be moved out of `diagnostic.rb`. Maybe `utils.rb` is a sensible place, probably even better a new file in `utils/`.

apjanke replied:
Yeah, that makes sense: you'd almost never actually want to do `/.*/` because there are so many small methods in most modules and classes, and if you do, it should be explicit. Removed the default and moved it to Utils.

UniqMartin commented on Library/Homebrew/diagnostic.rb:
```diff
@@ -1435,5 +1420,34 @@ def all
+      end
+    end
+
+    if $times.nil?
```
We don't expect `inject_dump_stats!` to be called multiple times in a single invocation, are we? Because of this, this check seems unnecessary.

apjanke replied:
Not right now, but why write code to be brittle like that when being robust is easy?

I added a check to make sure that methods are not double-instrumented, either.

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -1234,4 +1224,9 @@ def audit_urls
+
+  if ARGV.switch? "D"
+    Homebrew.inject_dump_stats!(FormulaAuditor, /^audit_*/)
+  end
```
This still needs to be moved into `def audit`. We don't want this code already to run for `require "cmd/audit"`.

apjanke replied:
Ah. I see why that matters now. Moved.

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -1234,4 +1224,9 @@ def audit_urls
+    Homebrew.inject_dump_stats!(FormulaAuditor, /^audit_*/)
+  end
 end
+
```
Superfluous empty line here. Sorry if I'm sounding pedantic, but looking at the diff prior to commit usually avoids this kind of diff noise.

apjanke replied:
Sorry; I've been addressing these feedback items individually without re-reading the whole diff on this. Will fix.

UniqMartin replied:
No worries. I guess I (or we) should give you some time to catch up with all the feedback. :smiley_cat:

apjanke commented:
Amended to address the latest round of output.

Two significant changes: 

Before, the `style` checks were only run when you specified files or formulae, and were skipped when you did a plain `brew audit` to check everything. Now it runs style checks too, since we can specify the correct set of files.

Removed the empty line between formulae in the output (and changed to 2-space indent for the bullet lists). Martin and I think it's more readable and a better use of space this way. IMHO, especially true now that `audit` has been tweaked to work better for large batches of files, and now includes `style` output in sections that previously did not have it.

New output:

```
$ brew audit a* b* c* --strict
homebrew/science/abinit:
  * Use '--without-test' instead of '--without-check'. Migrate '--without-check' with `deprecated_option`.
homebrew/science/abyss:
  * Options should begin with with/without. Migrate '--enable-maxk=' with `deprecated_option`.
  * Use '--without-test' instead of '--without-check'. Migrate '--without-check' with `deprecated_option`.
homebrew/science/adam:
  * `option` (line 21) should be put before `depends_on` (line 14)
  * Description is too long. "name: desc" should be less than 80 characters.
Length is calculated as adam + desc. (currently 112)
  * Use '--without-test' instead of '--without-check'. Migrate '--without-check' with `deprecated_option`.
homebrew/science/alembic:
  * `head` (line 17) should be put before `bottle block` (line 10)
  * A `test do` test block should be added
homebrew/science/alglib:
  * Formula should have a desc (Description).
homebrew/science/alpscore:
  * Use '--with-test' instead of '--with-check'. Migrate '--with-check' with `deprecated_option`.
  * C: 27: col 25: Don't use parentheses around a method call.
homebrew/science/amos:
  * Description shouldn't start with an indefinite article (A)
```

The format for multi-line problems is ugly, with 2nd and later lines not being indented, like "Length..." in the following.. Going to do something about that while I'm in here.

```
  * Description is too long. "name: desc" should be less than 80 characters.
Length is calculated as adam + desc. (currently 112)
  * Use '--without-test' instead of '--without-check'. Migrate '--without-check' with `deprecated_option`.
```

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -152,11 +142,21 @@ def initialize(formula, options = {})
+    return unless @style_offenses
+    display_cop_names = ARGV.include?("--display-cop-names")
+    @style_offenses.each do
+      |offense| problem offense.to_s(:display_cop_name => display_cop_names)
```
`|offense|` should be at the end of the preceding line, not at the start of this line.

apjanke replied:
Fixed.

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -1234,4 +1235,5 @@ def audit_urls
   def problem(text)
     @problems << text
   end
+
```
And here continues my crusade against spurious changes. :wink:

apjanke replied:
Okay, okay, amended. Happy now? :)

UniqMartin replied:
Happy. :grinning: I hope I'm not driving you nuts with my pedantry …

apjanke replied:
Not at all. This is stuff I care about too; it's just slipping past me in my edit / amend / comment / push cycle. I need to figure out a better work flow for this.

UniqMartin commented on Library/Homebrew/utils.rb:
```diff
@@ -276,6 +276,39 @@ def self.install_gem_setup_path!(gem, version = nil, executable = gem)
+      $times = {}
+      at_exit {
+        col_width = [$times.keys.map(&:size).max + 2, 15].max
+        puts $times.sort_by { |_k, v| v }.map {
```
I know this is legacy code, but multi-line blocks should use `do`-`end` (same problem also two lines above). Also the block arguments are again misplaced on the next line.

apjanke replied:
I pulled the block arguments up to the first line, and did the `do ... end` for the `at_exit` block.

But when I switch this particular `{ ... }` to a `do ... end`, the behavior changes. Instead of printing a list of lines

```
Error: 4 problems in 3 formulae
audit_github_repository:    0.0000 sec
audit_style:                0.0000 sec
audit_bottle_spec:          0.0000 sec
audit_conflicts:            0.0000 sec
```

it does a stringification of an Enumerator like this.

```
Error: 4 problems in 3 formulae
#<Enumerator:0x007fb78c6c0d68>
```

Know what's going on there? `ruby` 2.0.0p643 on OS X 10.9.5.

UniqMartin replied:
This must be the result of different operator precedence. `{ }` has precedence different from `do end`. If in doubt, saving an intermediate result in a variable is not the worst solution. Or in this particular instance you should be able to replace `sort_by { |_k, v| v }` with `sort_by(&:last)` to address the issue.

apjanke commented:
That turned out to be easier than I feared.

```
homebrew/science/adam:
  * `option` (line 21) should be put before `depends_on` (line 14)
  * Description is too long. "name: desc" should be less than 80 characters.
    Length is calculated as adam + desc. (currently 112)
  * Use '--without-test' instead of '--without-check'. Migrate '--without-check' with `deprecated_option`.
homebrew/science/alembic:
```

I think this is ready to go now, pending y'all's feedback on the output format change.

UniqMartin commented on Library/Homebrew/utils.rb:
```diff
@@ -276,6 +276,39 @@ def self.install_gem_setup_path!(gem, version = nil, executable = gem)
+      $times = {}
+      at_exit do
+        col_width = [$times.keys.map(&:size).max + 2, 15].max
+        puts $times.sort_by { |_k, v| v }.map { |method, time|
```
`do`-`end` for multi-line blocks please.

apjanke replied:
Same problem as before - if I change this `{ ... }` to `do ... end`, it will print "`Enumerator<#...>`" instead of displaying the individual lines.

Maybe it's a weird parsing issue related to precedence and the fact that there are two blocks introduced in this statement?

dunn replied:
I think it's trying to `map` the result of `puts` instead of vice-versa (http://stackoverflow.com/a/5513622); I guess you could add parentheses like `($times.sort_by ... )` or just stick with the braces.

UniqMartin replied:
Sorry for the nagging! Hadn't seen your other related comment. :see_no_evil:

UniqMartin replied:
Actually, the most sensible solution to this style issue seems to be to move the `puts` into the block (and to replace `map` with `each`):

```ruby
$times.sort_by { |_k, v| v }.each do |method, time|
  puts format("%-*s %0.4f sec", col_width, "#{method}:", time)
end
```

apjanke replied:
I already tried sticking parentheses in a couple places there, but didn't find an arrangement that worked.

Switched to Martin's suggestion.

UniqMartin commented on Library/Homebrew/utils.rb:
```diff
@@ -276,6 +276,39 @@ def self.install_gem_setup_path!(gem, version = nil, executable = gem)
+      at_exit do
+        col_width = [$times.keys.map(&:size).max + 2, 15].max
+        puts $times.sort_by { |_k, v| v }.map { |method, time|
+          "%-#{col_width}s %0.4f sec" % ["#{method}:", time]
```
The correct way to get `col_width` into the format string is (note that `format` is also to be preferred over `String#%`, according to RuboCop):

```ruby
format("%-*s %0.4f sec", col_width, "#{method}:", time)
```

apjanke replied:
Changed to use this.

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -1,87 +1,77 @@
 require "cmd/search"
+require "cmd/style"
 require "date"
+require "diagnostic"
```
I guess this can be removed now that `inject_dump_stats!` lives somewhere else.

apjanke replied:
Done

UniqMartin commented on Library/Homebrew/cmd/audit.rb:
```diff
@@ -1,87 +1,76 @@
 #:
-#:    If `--strict` is passed, additional checks are run. This should be used
-#:    when creating for new formulae.
+#:    If `--strict` is passed, additional checks are run, including Rubocop
```
Alright, switching into super-pedantic mode: The project name is “Rubo**C**op”, not “Rubocop”. (This is just one of several instances.)

apjanke replied:
Well, I'll be a monkey's uncle. IMDB says the movie is "RoboCop", too. Fixed.

I left capitalization of class names as `RubocopResults` and the like, following the Java CamelCase convention that internal capitalization styling is ignored for single words and acronyms, e.g. `JsonParser`, not `JSONParser`.

UniqMartin commented on Library/Homebrew/cmd/style.rb:
```diff
@@ -1,3 +1,25 @@
+#:    style checks on the whole Homebrew `Library`, including core code and all
+#:    formulae.
+#:
+#:    If `--fix` is passed and `$HOMEBREW_DEVELOPER` is set, style violations
```
I don't think we include the dollar sign in other parts of our man page when referring to environment variables.

apjanke replied:
Looks like we don't. Fixed.

mikemcquaid commented:
Looking good to me 👍. We can always iterate after shipping.

UniqMartin commented:
:+1: I can think of a few minor improvements, but they can wait and I've pestered you enough in this PR. I'm really happy with the code now (and the major improvement to `brew audit` hidden therein).

What I'd really like you to consider for future PRs (or even this PR, if you feel like retroactively doing this) would be not to lump all changes into a single commit. It's really hard to digest once some time has passed and you're looking at the Git history. Looking at the current code, here's what the theme of individual commits could be (they should be all independent and build on top of each other):

- utils: add `popen_read_text`
- utils: refactor `inject_dump_stats!` to be independent of `Homebrew::Diagnostic::Checks` (and adapt `cmd/doctor.rb`, `diagnostic.rb` accordingly).
- style: refactor and use RuboCop JSON interface
- style: add documentation
- audit: use generalized `inject_dump_stats!`
- audit: better integrate style checks

apjanke commented:
Sorry! I've gotten used to squashing to a single commit to formula changes; maybe overapplied it here.

I could still break these out to multiple commits if you think it would help readability. (Some `git` clients I use make that pretty easy; maybe they foresaw this situation.)

UniqMartin commented:
> Sorry! I've gotten used to squashing to a single commit to formula changes; maybe overapplied it here.

No need to apologize! It's just something that I noticed and I wanted to make you aware of. Just pretend for a moment that there's a breaking change in `audit` (very unlikely) introduced by this code. It would be a shame to have to revert all the other improvements also contained in the same commit and then again work on a fairly big chunk to get things sorted out.

> I could still break these out to multiple commits if you think it would help readability. (Some `git` clients I use make that pretty easy; maybe they foresaw this situation.)

I'd leave this decision up to you. It's obviously easier to do when things get split right from the start. I'm also fine merging it as-is, as nobody except me complained and even I did so at the very last moment.

apjanke commented:
Well, I was wrong: GitUp makes it easy to split commits on a per-file basis, and SourceTree has good support for staging/unstaging by line or hunk, but neither is going to help much with splitting this

apjanke commented:
Turns out splitting this would take a good amount of work: I'd want to make sure each individual commit still worked, so it would take manual editing and several rounds of testing, and I'm short on time this week. Let's just merge as-is, and I'll work on organizing future PRs better.

apjanke commented:
Shipped. Thanks for your feedback, folks!
